experimental/casemap/src/casemapper.rs

```diff
@@ -142,20 +146,63 @@ impl CaseMapper {
     /// as a `LanguageIdentifier` (usually the `id` field of the `Locale`) if available, or
     /// `Default::default()` for the root locale.
     ///
-    /// See [`Self::titlecase_segment_to_string()`] for the equivalent convenience function that returns a String,
+    /// This function performs legacy head adjustment behavior when [`HeadAdjustment::Adjust`] is set. See
+    /// the docs of [`TitlecaseMapper`] for more information on what this means. There is no difference between
+    /// the behavior of this function and the equivalent ones on [`TitlecaseMapper`] when the head adjustment mode
+    /// is [`HeadAdjustment::NoAdjust`].
+    ///
+    /// See [`Self::titlecase_segment_legacy_to_string()`] for the equivalent convenience function that returns a String,
     /// as well as for an example.
-    pub fn titlecase_segment<'a>(
+    ///
+    /// [`TitlecaseMapper`]: crate::TitlecaseMapper
+    pub fn titlecase_segment_legacy<'a>(
         &'a self,
         src: &'a str,
         langid: &LanguageIdentifier,
+        options: TitlecaseOptions,
     ) -> impl Writeable + 'a {
-        self.data.get().full_helper_writeable::<true>(
-            src,
+        self.titlecase_segment_with_adjustment(src, langid, options, |data, ch| data.is_cased(ch))
+    }
+
+    /// Helper to support different head adjustment behaviors,
+    /// `char_is_head` is a function that returns true for a character that is allowed to be the
+    /// titlecase head, when `head_adjustment = Adjust`
+    ///
+    /// We return a concrete type instead of `impl Trait` so the return value can be mixed with that of other calls
+    /// to this function with different closures
+    pub(crate) fn titlecase_segment_with_adjustment<'a>(
+        &'a self,
+        src: &'a str,
+        langid: &LanguageIdentifier,
+        options: TitlecaseOptions,
+        char_is_head: impl Fn(&CaseMapV1, char) -> bool,
+    ) -> StringAndWriteable<FullCaseWriteable<'a, true>> {
+        let data = self.data.get();
+        let (head, rest) = match options.head_adjustment {
+            HeadAdjustment::Adjust => {
+                let first_cased = src.char_indices().find(|(_i, ch)| char_is_head(data, *ch));
+                if let Some((first_cased, _ch)) = first_cased {
+                    (
+                        src.get(..first_cased).unwrap_or(""),
+                        src.get(first_cased..).unwrap_or(""),
+                    )
```

Comment:
Thought: It would be nice if `slice` had a function returning the signature `(&[T], &[T])` split at a predicate function